### ISSUE TITLE
Support for cleartext HTTP/2 (h2c)

### ISSUE THREAD

**Issue by vfaronov** [mitmproxy/mitmproxy#2116]
##### Steps to reproduce the problem:

1. Start listening on local port 8081 like this: `nc -vl 8081`
2. Start mitmproxy like this: `mitmdump --http2 --reverse http://localhost:8081`
3. Use [nghttp](https://nghttp2.org/documentation/nghttp.1.html) to make a request like this: `nghttp -v http://localhost:8080/`
4. Port 8081 receives the following:

```
PRI * HTTP/2.0
:authority: localhost
host: localhost:8081
```

This looks like the [HTTP/2 connection preface](https://tools.ietf.org/html/rfc7540#section-3.5) interpreted by mitmproxy (?) as an actual HTTP/1.1 request.

##### Any other comments? What have you tried so far?

The same happens if I use [Hyper](http://hyper.readthedocs.io/en/latest/CLI.html) in direct mode (`hyper --h2 GET http://localhost:8080/`) instead of nghttp.

I would expect mitmproxy to forward nghttp’s request to port 8081 in direct cleartext HTTP/2, though cleartext HTTP/1.1 and upgrading cleartext HTTP/2 would also be understandable.

See also #2115.

##### System information

(this is mitmproxy master)

Mitmproxy version: 3.0.0 (release version)
Python version: 3.5.2
Platform: Linux-4.4.0-65-generic-x86_64-with-Ubuntu-16.04-xenial
SSL version: OpenSSL 1.0.2g 1 Mar 2016
Linux distro: Ubuntu 16.04 xenial

**Comment by vfaronov** [comment]
> But I think either way, the client should first send an Upgrade-request

Not necessarily: see [RFC 7540 § 3.4](https://tools.ietf.org/html/rfc7540#section-3.4).

**Comment by Kriechi** [comment]
Ok - yet another feature that is not implemented yet 😄 
We are happy to accept PRs!

**Comment by mhils** [comment]
Supporting cleartext h2 would introduce quite a bit of complexity, so I'm actually 👎 to this. We should rather provide users with the option to enforce a h2 ALPN (see https://discourse.mitmproxy.org/t/using-mitmproxy-to-terminate-tls-for-http-2/404), then one could use two mitmproxy instances to work around this.

TL;DR: I'm voting to close this in favor of adding the option to enforce a specific ALPN. @Kriechi, what are your thoughts?

**Comment by Kriechi** [comment]
👍

**Comment by vfaronov** [comment]
@mhils 

> then one could use two mitmproxy instances to work around this

Can you elaborate on what you mean by this?

**Comment by vfaronov** [comment]
No TLS is involved in this issue. It’s h2c in, h2c out. So I don’t think ALPN can help here.

But this isn’t an actual problem for me; I was only trying stuff out.

**Comment by mhils** [comment]
> No TLS is involved in this issue. It’s h2c in, h2c out. So I don’t think ALPN can help here.

True - got confused here, sorry. Anyways, looks like we all agree that we don't want to support h2c-h2c right now.
